### PR TITLE
Block List: Display inserter button after block on hover, focus

### LINKED ISSUE #2752: Block List: Display inserter button after block on hover, focus
Related: #2743, #833

With an exploration of nested blocks pending (#2743), the usability of the global block inserter is being called into question. As an example, consider the case that the quote block allows nesting text or image blocks. If while focused within the quote text I decide to use the global inserter to insert an image block, would I expect it to (a) be inserted within the quote or (b) be inserted as a new top-level block after the quote? While the insertion point highlight helps, it is not shown until after the user has made the conscious decision to use the inserter (assuming they had already decided the effects of said insertion).

__Proposal:__

A separate task will be created to eliminate the global inserter.

We should first prepare for this by establishing an alternative behavior for inserting before/after/between blocks. When hovering or focusing a block, an inserter button should be made visible after that block to allow a new block to be inserted between it and the next.

Complications:

- An hover inserter should be shown prior to the first block. It is unclear how this could work with focus and tab order (maybe it doesn't)
- An inserter should always remain visible at the end of a post, so either the last block's hover inserter is not rendered or assumes the role of what currently exists as the always-visible inserter

cc @adambbecker (I am told you have explored this previously 😄 )

### PR DESCRIPTION
**Pull request by aduth**
Closes #2752
Related: #833

This pull request seeks to display a block inserter after a hovered or focused block, enabling a user to easily insert between blocks. This is a first step toward eventually dropping the header global inserter (#2755).

image

__Design notes:__

This is a best-attempt design. Iterations or suggestions are welcome. In particular, the specifics of the clickable/tappable area and hover bounds may need some improvement.

__Testing instructions:__

Verify that you can insert after a block which is focused or hovered, and that regressions do not otherwise occur with the global inserter or end-of-content inserters.

1. Navigate to Gutenberg > New Post
2. Insert a paragraph
3. Note that hovering this paragraph exposes no inserter, since it's the last block (expected to use end-of-content inserter)
4. Insert a second block
5. Hover and/or focus the first paragraph inserted in step 2
6. Note that an inserter is shown, enabling you to insert a new block after the paragraph

### CODE REVIEW

File: editor/inserter/index.js
Comment (youknowriad) [inline editor/inserter/index.js]: When is this second check necessary?

File: editor/inserter/menu.js
Comment (youknowriad) [inline editor/inserter/menu.js]: Could you explain why this change is necessary?

File: editor/modes/visual-editor/block.js
Comment (youknowriad) [inline editor/modes/visual-editor/block.js]: I guess this check is to avoid "removing" the inserter when we hover out of the block? 
Maybe we can handle this in a simpler way: select the block when we open its inserter.
I'm just trying to avoid this dependency between the inserter and this component.

**Review by youknowriad — COMMENTED** [review]
Great job so far.

I noticed a small "glitch" which is not totally related but I think we this change, we have to update the inserter at the bottom of the editor to always insert at the end of the post instead of after the selected block.

**Comment by gziolo** [comment]
@aduth, thank you for working on this one. In my opinion, this was the biggest UX issues I discovered so far in Gutenberg 🥇

File: editor/inserter/index.js
Comment (aduth) [reply editor/inserter/index.js]: The global and end-of-content inserters don't specify the `insertIndex` prop, so this ensures the sibling insertion point is only set for the between-inserters. It _might_ be the case that `insertIndex !== undefined` could work well here. I used `Number.isInteger` elsewhere to capture `0` but not `null` (since JavaScript is a very sensible language where `null >= 0` evaluates to true 😆 )

File: editor/inserter/menu.js
Comment (aduth) [reply editor/inserter/menu.js]: This is a similar issue to the one discussed at #2706: Namely, WritingFlow is handling the arrow presses and moving between blocks. The `true` argument allows us to handle the key press earlier during the capture phase. I might take another pass at this, and I wonder why we need to bind to the document at all instead of using React events (since I expect bubbling would work better in the latter case). If I keep it as-is, I'll add a clarifying comment.

File: editor/modes/visual-editor/block.js
Comment (aduth) [reply editor/modes/visual-editor/block.js]: >Maybe we can handle this in a simpler way: select the block when we open its inserter.

That sounds like a good idea! I'll give it a try.

**Comment by aduth** [comment]
>I noticed a small "glitch" which is not totally related but I think we this change, we have to update the inserter at the bottom of the editor to always insert at the end of the post instead of after the selected block.

Yep, makes sense. I'll change this.

**Comment by gziolo** [comment]
I tested this PR. I love the changes introduced ❤️  

Just saying that this toolbar doesn't play well with this change. I think that the toolbar needs some rethinking, because it also makes it harder to read the previous paragraph /cc @jasmussen

screen shot 2017-10-09 at 10 27 20

The last block looks a bit different, but I think it makes a lot of sense:

screen shot 2017-10-09 at 10 30 54

I can confirm that the last commit fixes issue mentioned by @youknowriad and the add code itself look good to me.

**Comment by gziolo** [comment]
@jasmussen what if we would move the very last + button to be rendered the same way as all other icons, but it would be always visible? We could also mirror the behavior of the existing + button and show on hover the following:

screen shot 2017-10-09 at 10 35 11

File: editor/inserter/index.js
Comment (youknowriad) [inline editor/inserter/index.js]: I guess we don't need this to be a class component, but probably the same discussion for/against inline props.

**Review by youknowriad — APPROVED** [review]
For me, this is good to ship. There are some UI concerns like pointed out by @gziolo but worth a try. Also, I'll explore the fixed toolbar switch soon which could solve the toolbar issue.

File: editor/inserter/index.js
Comment (gziolo) [reply editor/inserter/index.js]: I hope we will see the day where there is only one supported way of creating components in React. I feel like some components are converted back and forth whenever requirements change.

**Comment by gziolo** [comment]
Yes, totally agree we should ship it as it is and continue the discussion here.

**Comment by jasmussen** [comment]
Designwise, I'd like to echo @gziolo's comments, this isn't ideal and it adds some visual load to the block silhuette. 

I don't want to block this from moving forward — nice work — but it does seem like a half-step. 

I understand the moving parts — it has to be accessible, it has to be fairly simple codewise, and the previous solution wasn't great either. But interface wise, we are at a point where we should be extremely careful to introduce visual regressions, especially since this is one of the critiques that is (fairly so) leveraged on Gutenberg these days.

One of the design patterns of Gutenberg is that it's possible to show further block options when a block is selected. A common editor pattern when you want to insert content in between paragraphs or other elements, is to make a linebreak. 

Perhaps we could show the inserter when a block is _selected_? Perhaps we could show an _empty line_ when a block is selected, and have the Enter button put the cursor there? Perhaps both? It seems like it is worth looking at how desktop publishing apps solve this.

Either way it's important we strike a balance between something that is easy to implement and solves the problem, with the cognitive load that it adds. In this case it feels like there's an imbalance. This implementation also doesn't solve the inserting images inline (#2043), and it seems like this inserter pattern would've been a good opportunity to look at that.

**Comment by gziolo** [comment]
As I mentioned above, not having such button available next to each block caused me the biggest struggle when I opened Gutenberg for the first time.

I must admit, I got to easily tempted to click Merge button in here :) I will be more cautious next time and wait for a wider feedback whenever we introduce UI/UX changes.

> Perhaps we could show the inserter when a block is selected? Perhaps we could show an empty line when a block is selected, and have the Enter button put the cursor there? Perhaps both? It seems like it is worth looking at how desktop publishing apps solve this.

At the moment when you hover over one of the existing blocks you can see arrows which allow moving the block up or down in the document hierarchy:

screen shot 2017-10-09 at 11 11 13

In that context add block button plays a similar function, but as you observed it adds another visual element to the block. Let's continue the discussion to #2925.

**Comment by mtias** [comment]
Adding a bit more context as I discussed this one with @aduth in person. The point was to add functionality that allowed the user to insert blocks in between other blocks easily and without hidden interactions.

Implementation wise I think we should treat this inserter as a separate element—or at least, visually separate. It should show when we hover the limit area between two blocks. Additionally, it can show up when a block is selected. Something more like this:

image

This is also in anticipation of nested blocks, which would use the pattern of the bottom inserter.

**Comment by aduth** [comment]
>Implementation wise I think we should treat this inserter as a separate element—or at least, visually separate. It should show when we hover the limit area between two blocks.

In trying to imagine how this would operate in practice, I get the sense that it might be frustrating to find _the sweet spot_ between hovering and individual block and hovering between blocks to reveal the inserter. It was already not-great in the implementation which was proposed here.

**Pull request merged**
